File: src/main/java/dragonbook/inter/expr/Rel.java
package dragonbook.inter.expr;

import dragonbook.lexer.Token;
import dragonbook.symbols.Array;
import dragonbook.symbols.Type;

public class Rel extends Logical {
    public Rel(Token tok, Expr x1, Expr x2) {
        super(tok, x1, x2);
    }

    public Type check(Type p1, Type p2) {
        if (p1 instanceof Array || p2 instanceof Array) return null;
        else if (p1 == p2) return Type.BOOL;
        else return null;
    }
}

File: src/main/kotlin/dragonbook/inter/expr/Rel.kt
package dragonbook.inter.expr

import dragonbook.error.ParseError
import dragonbook.lexer.Token
import dragonbook.symbols.Array

data class Rel(
    override val op: Token,
    override val expr1: Expr,
    override val expr2: Expr
) : Logical {

    init {
        validate()
    }

    override fun validate() {
        if (expr1.type is Array || expr2.type is Array || expr1.type != expr2.type) throw ParseError(node = this)
    }

    override fun toString(): String = "$expr1 $op $expr2"
}